src/multiformats/address.cljc

```diff
@@ -2,29 +2,232 @@
+        code-bytes (varint/encode code)
+        val-bytes (if (nil? value)
+                    (b/byte-array 0)
+                    (codec/str->bytes codec value))
+        val-len-bytes (if (codec/fixed-byte-length codec)
+                        (b/byte-array 0)
+                        (varint/encode (alength ^bytes val-bytes)))]
+    (concat-arrs code-bytes val-len-bytes val-bytes)))
+
+(deftype Address
+  [^bytes _data _meta ^:unsynchronized-mutable _hash]
+
+  #?(:clj Seqable :cljs ISeqable)
+  (#?(:clj seq :cljs -seq) [this] (protocol-value-seq _data))
+
+  #?(:clj IPersistentCollection)
+
+  #?(:cljs ICounted)
+  (#?(:clj count :cljs -count)
+    [this] (count (seq this)))
+
+  #?(:cljs IEmptyableCollection)
+  (#?(:clj empty :cljs -empty)
+    [this] (Address. (b/byte-array 0) nil 0))
+
+  #?(:cljs IEquiv)
+  (#?(:clj equiv :cljs -equiv) [this other]
+    (and (instance? Address other)
+         (b/bytes= (.-_data this) (.-_data ^Address other))))
+
+  #?(:cljs ICollection)
+  (#?(:clj cons :cljs -conj) [this entry]
+    (Address. (concat-arrs _data (entry-bytes entry)) _meta 0))
+
+  IMeta
+  (#?(:clj meta, :cljs -meta)
+    [this]
+    _meta)
+
+
+  #?(:clj IObj :cljs IWithMeta)
+  (#?(:clj withMeta :cljs -with-meta)
+    [this new-meta]
+    (Address. _data new-meta _hash))
+
+  #?(:cljs IHash)
+  (#?(:clj hashCode, :cljs -hash)
+    [this]
+    (if (zero? _hash)
+      (let [hc (hash (b/byte-seq _data))]
+        (set! _hash hc)
+        hc)
+      _hash))
+
+  Object
+  (toString [this]
+     (->> this
+          (mapcat (fn [[k v]] [(name k) v]))
+          (remove nil?)
+          (str/join "/")
+          (str "/"))))
+
+
+(defn- parse-entries
+  [address-str]
+  (let [parts (->> #"/" (str/split address-str) (remove str/blank?))]
+    (loop [parts parts pairs []]
+      (if-let [protocol-key (-> parts first keyword)]
+        (let [{:keys [code, codec] :as attrs}
+              (ensure-protocol-attrs protocol-key)
+              fixed-len (codec/fixed-byte-length codec)]
+          (if (and fixed-len (zero? fixed-len))
+            (recur (next parts) (conj pairs [protocol-key nil]))
+            (let [new-entry [protocol-key (str (second parts))]]
+              (recur (drop 2 parts) (conj pairs new-entry)))))
+        pairs))))
+
+
+(defn parse
+  "parse a human-readable multiaddr string; the data that backs
+   the result is a packed representation.
+
+  The `Address` result can be treated as a collection over
+  the [protocol-key value] pairs that you can seq over
+  or directly `(conj addr [key val])` pairs unto
+  "
+  ^Address [address-str]
+  (let [pairs (parse-entries address-str)
+        pair-bytes (map entry-bytes pairs)]
+    (->Address (apply concat-arrs pair-bytes) nil 0)))
+
+(defn create
+  "create address from entries (possibly empty) you can build on
+
+  (str (create [[:ip4 \"127.0.0.1\"][:tcp \"80\"]]))
+  >> \"/ip4/127.0.0.1/tcp/80\" "
+  [& entries]
+  (into (->Address (b/byte-array 0) nil 0) entries))
```

Comment:
A more efficient approach here would be to directly encode the entries into a byte array before constructing the Address type. Without special reduction protocols, it seems like this is going to allocate O(n^2) memory.

Reply:
Done, now this just calls `(map entry-bytes)` and then concatenates results so memory is now O(n)